File: sorting_algos/insertion_sort.py
# Iterative Insertion Sort
def iterative_insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


# Recursive Insertion Sort
def recursive_insertion_sort(arr):
    def insertion_sort_recursive(arr, n):
        if n <= 1:
            return
        insertion_sort_recursive(arr, n - 1)
        key = arr[n - 1]
        j = n - 2
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

    insertion_sort_recursive(arr, len(arr))

File: seminar_1/task_1/InsertionSort.java
package algos_data_structures.seminar_1.task_1;

public class InsertionSort {
    // Recursive InsertionSort
    public static void insertionSortRecursive(int[] arr, int n) {
        if (n <= 1) return;
        insertionSortRecursive(arr, n - 1);
        int last = arr[n - 1];
        int j = n - 2;
        while (j >= 0 && arr[j] > last) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = last;
    }

    // Iterative InsertionSort
    public static void insertionSortIterative(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }
            arr[j + 1] = key;
        }
    }
}

